[1단계 - 콘솔 기반 로또 게임] 데이지(김유경) 미션 제출합니다.

kimyou1102 opened the pull request:
안녕하세요 쵸파! 데이지입니다. 이번 페어는 3인(다이앤😊, 세라🍀)으로 진행했어요. 페어들과 제가 고민한 것들을  작성해 두었습니다. 잘부탁드립니다!!

## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [ ] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: 

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 

1.
저희는 역할에 따라 다음과 같이 구분했습니다. 
각 class에 역할이 잘 분배 되어 있는지
`Lotto` : 숫자 확인 로직들
`LottoResult` : 당첨내역/수익률 계산
`LottoMachine` : 발행된 로또들 저장, 발행한 로또들을 돌면서 숫자 확인하고 순위 반환
`Ticket`: 로또 발행 객체

해당하는 역할을 담당하는 도메인들을 나누는 것에 고민을 많이 했어요. 해당 구조가 각 도메인에 대해 관심사가 잘 나누어졌는지 궁금합니다!

2. 검증에 대해 더 확장성있게 만들고 싶어서 공통이 되는 Validator을 만들고 해당 클래스를 상속 받아서 LottoValidator, PriceValidator 식으로 만들어서 사용했어요. 추후 명세가 바뀌었을때 해당 구조가 잘 대응할 수 있는 구조일까요? 쵸파의 의견이 궁금해요!

### 2) 이번 리뷰를 통해 논의하고 싶은 부분
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

1. TDD를 통해 작은 단위로 문제를 바라볼 수 잇다는 것은 좋은 점이라고 느꼈어요. 하지만 도메인에 대해 숙지 되지 않은 상태에서 TDD를 사용하다 보니 구조를 생각하다 보니 시간이 더 오래 걸리고 어느 장점이 있는지 와닿지 않는 것 같아요. 기능 개발 후 테스트를 작성하는 것과의 뚜렷한 장점도 잘 느끼지 못한 것 같아요. 쵸파는 TDD의 장점이 뭐라고 생각하시는지 궁금합니다!
2. 페어 프로그래밍을 하면서 메소드 선언 순서에 대해 의견이 갈렸어요. 저는 코드의 가독성을 높이기 위해, 중요한 메서드 순으로 메서드를 정렬해야 한다고 생각했어요. 왜냐하면 핵심이 되는 로직을 상단에 배치할 수록 다른 사람이 그 파일에 대해 파악하기 쉽다고 생각했습니다. 쵸파는 어떤 기준으로 메서드를 선언하는 것을 선호하시는지 궁금합니다!

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [ ] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

keemsebin committed:
chore: 기능목록 readme 작성

Co-Authored-By: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-Authored-By: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 입력받은 금액 만큼 로또 개수 생성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 정해진 개수만큼 로또 생성 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: random 값 생성 함수 구현, 중복되지 않은 랜덤 값을 담은 배열 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: Lotto 번호 주입, 로또 번호 get 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 생성 시 랜덤 번호 주입 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 번호 오름차순 정렬 메소드 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
test: 로또 번호 오름차순 테스트 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 기본 입력 함수 구현 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 랜덤으로 주어진 로또 번호와 입력받은 로또 번호의 일치 갯수 함수 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 로또에 보너스 번호 존재 여부 확인 함수 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 발행한 로또 번호와 입력한 로또 번호/보너스 번호의 일치 갯수를 기반으로 로또 상태를 매칭하는 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 당첨내역을 기반으로 총 수입을 구하는 메소드 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 당첨 내역을 기반으로 수익률을 계산하는 메소드 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
fix: createLotto 빈배열 반환하는 문제 해결

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
fix: getMatchedLottoRank 함수 매개변수에 bonusNumber 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 구입 금액 입력 로직 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: INPUT, OUTPUT 상수화

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: index.js에 당첨번호, 보너스 번호, 재입력 로직 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
chore: readme 유효성 검사 케이스 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 출력 기능을 담당하는 객체 생성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: Validator class 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 에러메세지 constants 정의

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 에러 throw utils 함수 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: Lotto class 내 validate 메소드 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: lotto validator class 생성, lotto validator test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 보너스 번호 validator class 구현, test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 구매한 금액에 대한 유효성 검사 class 생성, test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 번호가 6개 미만인 경우 validation 조건 추가, test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 전체 로직 index.js에 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
fix: printLottoNumber 출력 시 lotto number 받아오는 함수 추가

keemsebin committed:
fix: retry 함수에 Input retry 객체를 다시 받아오도록 수정

keemsebin committed:
refactor: createLotto 함수를 Ticket 객체 매소드로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: count 함수명 변경

keemsebin committed:
chore: 사용하지 않는 import 제거

keemsebin committed:
feat: 상수 Object freeze 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: 일치하는 로또 상태를 반환하는 메소드명 변경

keemsebin committed:
refactor: updateFinalStatus 메소드 추가, getMatchedLottoStatus 메소드 간소화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 당첨 통계, 라인 출력 문자 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: Ticket 객체에 로또 발행 함수 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: index.js Input 로직 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
chore: readme 폴더 구조 추가

keemsebin committed:
fix: printInput 함수 화살표 함수 누락 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

bassyu commented on src/constants/message.js:
```diff
@@ -0,0 +1,27 @@
+export const ERROR = Object.freeze({
+  EMPTY: "빈 값은 입력할 수 없습니다.",
+  NOT_NUMBER: "숫자가 아닌 값은 입력할 수 없습니다.",
+  INVALID_RANGE: "범위를 벗어난 입력은 할 수 없습니다.",
```
허용범위가 어느정도인지 정확하게 알려주면 좋을 것 같아요!

kimyou1102 replied:
에러 메시지를 더 자세히 적으면 사용자가 어떻게 행동을 수정하면 될지 알기 쉽겠네요 반영하겠습니다!

bassyu commented on src/domain/Lotto.js:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  sortLottoNumber(numbers) {
+    return numbers.sort((a, b) => a - b);
```
지금은 문제 없지만, 원본 배열을 수정하는 함수는 항상 유의하시길 바랍니다!

kimyou1102 replied:
생성자에서 바로 넘겨주는 거라 크게 신경쓰지 않았는데 원본 배열을 수정하는 것은 유의해야 하는 것이니 복사값을 변경해주는 것으로 변경하겠습니다!

bassyu commented on src/constants/lotto.js:
```diff
@@ -0,0 +1,13 @@
+  LENGTH: 6,
+});
+
+export const LOTTO_STATUS = Object.freeze([
+  { RANK: 1, COUNT: 6, REWORD: 2_000_000_000, IS_BONUS: false },
+  { RANK: 2, COUNT: 5, REWORD: 30_000_000, IS_BONUS: true },
+  { RANK: 3, COUNT: 5, REWORD: 1_500_000, IS_BONUS: false },
+  { RANK: 4, COUNT: 4, REWORD: 50_000, IS_BONUS: false },
+  { RANK: 5, COUNT: 3, REWORD: 5000, IS_BONUS: false },
+]);
```
이렇게 그룹화 해주신 것 너무 좋네요! 👍👍

bassyu commented on src/domain/LottoResult.js:
```diff
@@ -0,0 +1,23 @@
+    this.winningHistory = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 };
+  }
+
+  getWinningHistory() {
```
이 메서드는 호출되는 횟수 만큼 history에 기록이 추가되는 사이드 이펙트가 있네요!
만약 누군가 이 메서드를 과하게 사용해서 버그가 생기면, 범인이 이 메서드라는 것을 찾기가 매우 힘들겠어요.

그래서 최대한 사이드 이펙트를 줄이거나, 
줄일 수 없다면 네이밍을 통해 표현하는 것이 중요하다고 생각합니다!

사이드 이펙트를 없애는 방향으로 수정할 수 있을까요?

kimyou1102 replied:
status를 업데이트 해주는 updateStatus메서드를 생성후 LottoResult객체의 생성자에서 호출. 그리고 getWinningHistory에서는 winningHistory를 뱉어주는 식으로 한다면 getWinningHistory에 대한 사이드 이펙트를 없애는 방향으로 갈 수 있을 거 같습니다!

```
  constructor(lottoStatus, price) {
    this.lottoStatus = lottoStatus;
    this.price = price;
    this.winningHistory = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 };

    this.updateWinningHistory();
  }

  updateWinningHistory() {
    this.lottoStatus.forEach((status) => {
      this.winningHistory[status.RANK] += 1;
    });
  }

  getWinningHistory() {
    return this.winningHistory;
  }
```

bassyu replied:
객체를 만들 때에만 계산하도록 수정해주셨군요! 
안전해진 것도 맞지만, 이해하기도 더 쉬워졌네요 👍

bassyu reviewed (changes requested):
안녕하세요 데이지~!
리뷰가 늦어서 정말 죄송합니다..! 🙇 😢

3인 페어는 2인 페어와 비교했을 때 어땠는지 궁금하네요~!

## 기능

잘 구현해주셨네요! 👍 👍 
다만 재시작이 딱 1번 가능한데, 의도된 동작이 맞는지 궁금합니다!

## 코드

거의 모든 메서드가 get~ 으로 시작하네요! 
그래서 단순히 가져오는 줄 알았는데, 계산을 하거나 구조를 변환하는 등 다양한 동작을 해주고 있어요.
그래서 함수 네이밍에 CRUD 보다는 훨씬 더 많은 의미를 담아줘야 할 것 같습니다!

> 해당하는 역할을 담당하는 도메인들을 나누는 것에 고민을 많이 했어요. 해당 구조가 각 도메인에 대해 관심사가 잘 나누어졌는지 궁금합니다!

잘 나눠주신 것 같아요! 
다만, 클래스를 사용하는 데이지만의 기준이 궁금합니다!

> 검증에 대해 더 확장성있게 만들고 싶어서 공통이 되는 Validator을 만들고 해당 클래스를 상속 받아서 LottoValidator, PriceValidator 식으로 만들어서 사용했어요. 추후 명세가 바뀌었을때 해당 구조가 잘 대응할 수 있는 구조일까요? 쵸파의 의견이 궁금해요!

공통된 부분을 따로 분리해주신 것은 좋습니다! 👍
그리고 특정 도메인의 검증이라면, 검증을 추가하거나 묶어서 각각 분리해주신 것도 아주 좋아보이네요~!

다만, 클래스의 메서드와 상속으로 구현해주신 이유가 궁금해요.
상속은 부모 클래스가 자식 클래스에게 '내가 변하면 너도 꼭 변해야해'라는 강제성 부여의 측면에서 이점이 있는데요, 
이 강제성으로 유연성을 손해본 부분도 있어보이네요!

예를 들어 LottoValidator와 PriceValidator에 필요한 `Validator`메서드가 추가된다면? BonusNumberValidator는 몰라도 되는 함수를 알게 되겠네요~!

## 논의

> TDD를 통해 작은 단위로 문제를 바라볼 수 잇다는 것은 좋은 점이라고 느꼈어요. 하지만 도메인에 대해 숙지 되지 않은 상태에서 TDD를 사용하다 보니 구조를 생각하다 보니 시간이 더 오래 걸리고 어느 장점이 있는지 와닿지 않는 것 같아요. 기능 개발 후 테스트를 작성하는 것과의 뚜렷한 장점도 잘 느끼지 못한 것 같아요. 쵸파는 TDD의 장점이 뭐라고 생각하시는지 궁금합니다!

데이지 말대로 오직 빠른 구현의 관점에서는 장점이 부족하다고 생각합니다!

안전한 코드의 관점에서 보자면, 오히려 느리게 구현할 수 있는 장점이 있다고 생각해요.
테스트를 먼저 고민하느라 뺏기는 시간을 대가로, 더 안전한 코드를 강제한다고 생각했습니다!

학습에 관점에서 보자면, 나중엔 구현부터 하더라도 지금 '함수를 테스트하기 좋은 단위로 분리'하는 것과 '다양한 엣지 케이스를 고려'하는 연습을 미리 해본 것이 도움된다고 생각해요~!

> 페어 프로그래밍을 하면서 메소드 선언 순서에 대해 의견이 갈렸어요. 저는 코드의 가독성을 높이기 위해, 중요한 메서드 순으로 메서드를 정렬해야 한다고 생각했어요. 왜냐하면 핵심이 되는 로직을 상단에 배치할 수록 다른 사람이 그 파일에 대해 파악하기 쉽다고 생각했습니다. 쵸파는 어떤 기준으로 메서드를 선언하는 것을 선호하시는지 궁금합니다!

취향이 많이 갈리는 주제네요~! '어제 잤어 나는', '잤어 어제 나는'처럼 순서가 달라도 충분히 이해는 되니까요...
저도 두괄식으로 말하면 이해가 잘 된다고 느껴서, 데이지 기준에 한표 하겠습니다~!

kimyou1102 committed:
fix: 재시작을 1번만 하는 오류 해결

- 재시도 응답이 y일 경우 start메서드를 호출하게 변경

kimyou1102 committed:
refactor: getWinningHistory 호출 시 발생하는 사이드 이펙트 제거

- getWinningHistory가 호출될 때마다 상태(winningHistory)를 업데이트하여 여러 번 호출 시 부작용 발생
- 상태 업데이트를 전담하는 updateWinningHistory 메서드를 생성하고, 생성자에서 한 번만 실행하도록 수정

kimyou1102 committed:
refactor: 원본값 수정이 아닌 복사한 값 수정으로 변경

kimyou1102 committed:
fix: 객체 상수의 요소가 아닌 객체를 넘겨준 문제 해결

kimyou1102 committed:
refactor: hasBonusNumber를 hasNumber로 변경하여 Lotto 객체의 역할에 맞게 수정

- hasBonusNumber는 특정한 보너스 번호 확인에 국한되지만, hasNumber는 더 범용적으로 사용할 수 있음
- Lotto 클래스가 로또 번호를 다루는 역할을 고려하여, 객체의 의미에 맞는 네이밍으로 변경

kimyou1102 committed:
refactor: 메서드명 명확하게 변경

- 같은 숫자들의 길이를 반환하는 메서드: getSameNumbers -> getSameNumbersLength
- 최종이익을 계산하여 반환하는 메서드: getTotalProfit -> calculateTotalProfit
- 수익률을 계산하여 반환하는 메서드: getRate -> calculateRate
- 모든 로또 상태를 업데이트해주는 메서드: updateFinalStatus -> updateAllLottoStatus

kimyou1102 committed:
refactor: Set의 중복 방지 특성을 활용하여 checkAndAddLottoNumbers 메서드 제거

kimyou1102 committed:
refactor: 메서드명을 역할에 맞게 명확하게 변경

- 로또 번호 생성 기능: issueLottoNumbers -> generateLottoNumbers
- 로또 객체 생성: createLottos -> issueLottos

kimyou1102 committed:
refactor: 객체명 객체의 역할이 잘 드러나도록 변경

- 로또를 발행하는 것들을 담은 객체: Ticket -> LottoFactory

kimyou1102 committed:
refactor: getMatchedLottoStatus의 역할을 getter로 순수하게 분리

- getMatchedLottoStatus에서 상태 변경 로직 제거하고, 오직 matchedLottoStatus 반환만 수행하도록 수정

kimyou1102 committed:
refactor: 도메인 클래스 속성에 대해 private 필드 적용

- Lotto, LottoMachine, LottoResult클래스에 적용

kimyou1102 committed:
test: 당첨된 내역이 없는 경우 수익률 테스트 추가

- test.each를 사용하여 다양한 당첨 내역에 대한 수익률 테스트를 추가

kimyou1102 committed:
test: Lotto 클래스의 테스트를 분리하여 파일 구조 정리

- utils 함수에 대한 테스트를 utils.test.js로 이동
- LottoFactory 클래스에 대한 테스트를 LottoFactory.test.js로 이동
- Lotto.test.js에는 Lotto 클래스 관련 테스트만 유지하여 역할 명확화

kimyou1102 committed:
test: 검증 테스트에서 경계값을 고려하여 테스트 데이터 변경

kimyou1102 commented:
안녕하세요 쵸파! 덕분에 생각하지 못한 부분들을 체크해 볼 수 있어서 좋았습니다! 피드백 반영해서 추가 리뷰 요청드리겠습니다😊

> 도메인 나눈거  잘 나눠주신 것 같아요! 다만, 클래스를 사용하는 데이지만의 기준이 궁금합니다!

객체를 여러개 만들어야 하는 상황이거나, 한 객체에서 속성을 공유해야 하는 경우, 상속이 필요한 경우를 기준으로 클래스를 사용했어요. 그래서 Ticket객체(LottoFactory)인 경우는 앞의 상황들이 아니기에 객체로 만들어서 사용했습니다.

> 다만, 클래스의 메서드와 상속으로 구현해주신 이유가 궁금해요.
상속은 부모 클래스가 자식 클래스에게 '내가 변하면 너도 꼭 변해야해'라는 강제성 부여의 측면에서 이점이 있는데요,
이 강제성으로 유연성을 손해본 부분도 있어보이네요!
예를 들어 LottoValidator와 PriceValidator에 필요한 Validator메서드가 추가된다면? BonusNumberValidator는 몰라도 되는 함수를 알게 되겠네요~!

공통되는 검증로직들을 Validator 클래스의 메서드로 만들고, 상속을 통해서 부모에서 받아오게 하면 검증 로직을 재사용하는 것에 이점이 있다고 생각했어요. 부모 클래스에 정말 공통되는 검증에 대한 메서드들을 선언해주고 그것들만 상속받아 자식 클래스를 받는다면 오히려 강한 결합이 이점일 거라 생각했습니다. 

하지만 쵸파가 적어주신 경우를 보고 오히려 명세가 변할수록 코드가 복잡해 질 수 있겠다는 생각이 들었습니다. 그래서 공통 검증 메서드를 모아둔 객체를 만들고 필요한 것만 빼서 사용하며, 특정 객체에서만 필요한 검증은 해당 객체에 적는 것이 오히려 유연성을 증가시킬 수 있는 방향이라 생각드네요!

bassyu commented on src/domain/LottoMachine.js:
LottoResult의 부수효과를 없앤 것처럼, 여기에도 부수효과를 제거할 수 있겠네요!
push를 통해 status를 수정하기 때문에 위험해보입니다!

그래서 무언가 수정할만한 것이 생기면, 혹시 다른곳도 그러지 않을까? 탐색해보는 것도 추천드립니다~!

bassyu reviewed (approved):
안녕하세요 데이지! 빠르게 반영해주셨네요!
2단계 진행해보면서 핑퐁 더 해봅시다~! 💪 

> 그래서 공통 검증 메서드를 모아둔 객체를 만들고 필요한 것만 빼서 사용하며, 특정 객체에서만 필요한 검증은 해당 객체에 적는 것이 오히려 유연성을 증가시킬 수 있는 방향이라 생각드네요!

넵ㅎㅎ 객체여도 되고, 그냥 함수로 풀어도 좋습니다!
강제성이 꼭 필요 없다면, 이 부분은 2단계에서 마저 반영해주세요~!
